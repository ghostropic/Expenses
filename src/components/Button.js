import React from 'react'
import styled from 'styled-components'

const Button = ({ text, onClick, disabled, className, type }) => (
  // eslint-disable-next-line react/button-has-type
  <button
    type={type}
    disabled={disabled}
    onClick={onClick}
    className={className}
  >
    {text}
  </button>
)

const StyledButton = styled(Button)`
  cursor: pointer;
  height: 48px;
  display: inline-block;
  padding: 0.5rem 0;
  width: 11rem;
  color: ${props => props.color || 'white'};
  border: 2px solid ${props => props.color || 'white'};
  background: ${props => props.background || 'transparent'};

  &.formButton {
    margin: 7px;
  }
  
  &.removeButton {
    background: palevioletred;
  }
 `

export default StyledButton
