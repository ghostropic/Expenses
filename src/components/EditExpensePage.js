import React from 'react'
import styled from 'styled-components'

const EditExpensePage = (props) => {
  console.log(props)
  return (
    <div>
      edit expenses for {props.match.params.id}
    </div>
  )
}

export default EditExpensePage