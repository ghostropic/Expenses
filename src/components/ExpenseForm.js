import React from 'react'
import moment from 'moment'
import {SingleDatePicker} from 'react-dates'

import 'react-dates/initialize'
import 'react-dates/lib/css/_datepicker.css'

export default class ExpenseForm extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? (props.expense.amount / 100) : '',
      note: props.expense ? props.expense.note : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      focused: false,
      error: ''
    }
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({createdAt}))
    }
  }

  onFocusChange = ({focused}) => {
    this.setState(() => ({focused}))
  }

  onDescriptionChange = (e) => {
    const description = e.target.value
    this.setState(() => ({description}))
  }

  onAmountChange = (e) => {
    const amount = e.target.value
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({amount})) 
    }
  }

  onNoteChange = (e) => {
    const note = e.target.value
    this.setState(() => ({note}))
  }

  OnSubmit = (e) => {
    e.preventDefault()
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({error: 'this is an error message'}))
      console.log('error')
    } else {
      this.setState(() => ({error: ''}))
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      })
    }
  }

  render() {
    return (
    <div> 
      <p>Expense Form:</p>
      <div>{this.state.error}</div>
      <form onSubmit={this.OnSubmit}>
        <input 
        value={this.state.description} 
        onChange={this.onDescriptionChange} 
        type="text" 
        placeholder="description" 
        autoFocus
        />
        <input 
        type="text" 
        placeholder="amount"
        value={this.state.amount}
        onChange={this.onAmountChange} 
        /> 
        <textarea 
        placeholder="Add a note to your expense."
        value={this.state.note}
        onChange={this.onNoteChange}
        ></textarea> 
        <SingleDatePicker
          date={this.state.createdAt} // momentPropTypes.momentObj or null
          onDateChange={this.onDateChange} // PropTypes.func.isRequired
          focused={this.state.focused} // PropTypes.bool
          onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
          id="expense_form_calender" // PropTypes.string.isRequired,
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <button>Add Expense</button>
      </form>
    </div>
    )
  }
}
