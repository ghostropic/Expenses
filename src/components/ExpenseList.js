import React from 'react'
import { connect } from 'react-redux'

import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'
import selectExpensesTotal from '../selectors/expenses-total'

export const ExpenseList = ({ expenses }) => (
  <div>
    <p>{`Total: ${selectExpensesTotal(expenses)}`}</p>
    <hr />
    {
      expenses.length === 0
        ? (<p>No Expense to list. Try clearing the date range.</p>)
        : (expenses.map(expense => <ExpenseListItem key={expense.id} {...expense} />))
    }
  </div>
)

const MapStateToProps = state => ({
  expenses: selectExpenses(state.expenses, state.filters)
})

export default connect(MapStateToProps)(ExpenseList)
