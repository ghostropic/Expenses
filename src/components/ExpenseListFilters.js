import React from 'react'
import {connect} from 'react-redux'
import * as actionFilters from '../actions/filters'

const ExpenseListFilters = (props) => (
  <div>
    <span>Filter:</span>
    <input type="text" value={props.filters.text} onChange={(e) => {
      props.dispatch(actionFilters.setTextFilter(e.target.value))
    }}/>
    <select value={props.filters.sortBy} onChange={(e) => {
      e.target.value === 'date' ?
        props.dispatch(actionFilters.sortByDate()) :
        props.dispatch(actionFilters.sortByAmount())
    }}>
      <option value="amount">Amount</option>
      <option value="date">Date</option>
    </select>
  </div>
)

const MapStateToProps = (state) => {
  return {
    filters: state.filters
  }
}

export default connect(MapStateToProps)(ExpenseListFilters) 