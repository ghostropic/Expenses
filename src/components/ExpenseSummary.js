import React from 'react'
import { connect } from 'react-redux';

import selectExpenses from '../selectors/expenses'
import selectExpensesTotal from '../selectors/expenses-total'

const toDollars = (pennies) => {
  const dollars = pennies / 100;
  return dollars.toLocaleString('en-US', { style: 'currency', currency: 'USD' });
}

export const ExpenseSummary = ({ expensesCount, expensesWord, expensesAmount }) => (
  <div>
    <p>{`Viewing ${expensesCount} ${expensesWord} totaling ${toDollars(expensesAmount)}`}</p>
    <hr />
  </div>
)

const MapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters)
  return {
    expensesCount: visibleExpenses.length,
    expensesAmount: selectExpensesTotal(visibleExpenses),
    expensesWord: visibleExpenses.length === 1 ? 'expense' : 'expenses'
  }
}

export default connect(MapStateToProps)(ExpenseSummary)
