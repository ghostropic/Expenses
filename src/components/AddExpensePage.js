import React from 'react'
import styled from 'styled-components'
import {connect} from 'react-redux'

import ExpenseForm from './ExpenseForm'
import expenses from '../reducers/expenses'
import { addExpense } from '../actions/expenses'

const AddExpensePage = (props) => (
  <div>
    add expense:
    <ExpenseForm
      onSubmit={(expense) => {
        props.dispatch(addExpense(expense))
        props.history.push('/')
      }}
    />
  </div>
)

export default connect()(AddExpensePage)