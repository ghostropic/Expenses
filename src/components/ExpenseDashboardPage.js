import React from 'react'
import styled from 'styled-components'
import { utils } from '../styles/style'

import ConnectedExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummary from './ExpenseSummary'

const ExpenseDashboardPage = () => (
  <Dashboard>
    <p>Search expenses.</p>
    <ExpenseListFilters />
    <ExpenseSummary />
    <ConnectedExpenseList />
  </Dashboard>
)

const Dashboard = styled.div`
  padding: ${utils.border};
  max-width: 800px;
`

export default ExpenseDashboardPage
