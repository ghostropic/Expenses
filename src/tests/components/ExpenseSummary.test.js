import React from 'react'
import { shallow } from 'enzyme'

import { ExpenseSummary } from '../../components/ExpenseSummary';

test('should render expense summary correctly with 0 expenses', () => {
  const wrapper = shallow(<ExpenseSummary expensesCount={0} expensesAmount={0} expensesWord="expenses" />)
  expect(wrapper).toMatchSnapshot()
})

test('should render expenses summary correctly with 1 expense', () => {
  const wrapper = shallow(<ExpenseSummary expensesCount={1} expensesAmount={100} expensesWord="expense" />)
  expect(wrapper).toMatchSnapshot()
})

test('should render expenses summary correctly with many expenses', () => {
  const wrapper = shallow(<ExpenseSummary expensesCount={10} expensesAmount={10000} expensesWord="expenses" />)
  expect(wrapper).toMatchSnapshot()
})
