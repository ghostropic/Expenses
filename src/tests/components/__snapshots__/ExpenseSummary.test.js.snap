// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense summary correctly with 0 expenses 1`] = `
<div>
  <p>
    Viewing 0 expenses totaling $0.00
  </p>
  <hr />
</div>
`;

exports[`should render expenses summary correctly with 1 expense 1`] = `
<div>
  <p>
    Viewing 1 expense totaling $1.00
  </p>
  <hr />
</div>
`;

exports[`should render expenses summary correctly with many expenses 1`] = `
<div>
  <p>
    Viewing 10 expenses totaling $100.00
  </p>
  <hr />
</div>
`;
